### PR TITLE
Throwing exception if single callable are not passed as List type 

### PR DESCRIPTION
**Pull request by rkscodes**
### Checklist

- [X] This pull request references any related issue by including "closes #9686"
- [ ] This pull request includes tests or only affects documentation.
- [X] This pull request includes a label categorizing the change `fix`.

### CODE REVIEW

File: src/prefect/flows.py [hunk 1]
```diff
@@ -158,6 +159,15 @@ def __init__(
...
         ] = None,
     ):
+        # Validate states is passed as list
+        states = [on_completion, on_failure, on_cancellation, on_crashed]
+        for state in states:
+            if state is not None and not isinstance(state, list):
+                raise TypeError(
+                    f"Expected list of callable for {state}; got"
+                    f" {type(state).__name__} instead."
+                )
```
Comment (madkinsz) [inline src/prefect/flows.py:169]: I'd call these `hook_types`/`hooks` instead of `states`/`state`. 

We probably want to support other iterables like tuples and sets here. Perhaps we can try to cast it to a list? e.g. `on_completion = list(on_completion)` with a try/except that captures the `TypeError` that is raised if a non-iterable type is given — we'd then raise it again with a nicer error?

This message you wrote also made me notice we could be checking if the items in the list are callables.

File: src/prefect/flows.py [hunk 2]
```diff
@@ -62,7 +62,8 @@
 from prefect.utilities.hashing import file_hash
 from prefect.utilities.importtools import import_object
 
-T = TypeVar("T")  # Generic type var for capturing the inner return type of async funcs
+# Generic type var for capturing the inner return type of async funcs
+T = TypeVar("T")
```
Comment (madkinsz) [inline src/prefect/flows.py:66]: Ideally we'd retain this to match the other type variables here

File: src/prefect/flows.py (hunk 1 above)
Comment (rkscodes) [reply src/prefect/flows.py:169]: Yeah sure, makes sense.

File: src/prefect/flows.py (hunk 2 above)
Comment (rkscodes) [reply src/prefect/flows.py:66]: I am sorry, I don't quite get this.

Comment (madkinsz) [reply src/prefect/flows.py:66]: Ah. You changed the position of this comment so it doesn't match the comments below it for the other type variables. This change isn't necessary for this pull request and you should undo it.

**Review by madkinsz — COMMENTED** [review]
Nice! This looks good to me. We'll want to add tests next in `test_flows.py` and `test_tasks.py` that use `with pytest.raises(...)` to ensure that the correct errors are raised.

File: src/prefect/flows.py [hunk 3]
```diff
@@ -158,6 +159,25 @@ def __init__(
             List[Callable[[schemas.core.Flow, FlowRun, State], None]]
         ] = None,
     ):
+        # Validate if hooks type is list and hook_type is callable
+        hooks = [on_completion, on_failure, on_cancellation, on_crashed]
+        for hook_types in hooks:
```
Comment (madkinsz) [inline src/prefect/flows.py:164]: Should probably swap use of `hooks` and `hook_types` here for clarity.

File: src/prefect/flows.py [hunk 4]
```diff
@@ -158,6 +159,25 @@ def __init__(
...
+            if hook_types is not None:
+                try:
+                    hook_types = list(hook_types)
+                except TypeError:
+                    raise TypeError(
+                        f"Expected list of callable for {hook_types}; got"
```
Comment (madkinsz) [inline src/prefect/flows.py:170]: This could probably say "Expected iterable for..." since we're covering callables in the next message

Comment (madkinsz) [reply src/prefect/flows.py:170]: We'll want to make sure this also points the user to the correct parameter i.e. "Expected iterable for 'on_cancellation'; got..." — right now it looks like it will display the value they passed.

**Comment by madkinsz** [comment]
It looks like you linked the wrong issue in the PR body as well.

File: src/prefect/flows.py (hunk 2 above)
Comment (rkscodes) [reply src/prefect/flows.py:66]: Hmm, I have no idea how it got changed, I think somehow black did it, anyways I am reverting it back.

File: src/prefect/flows.py (hunk 3 above)
Comment (rkscodes) [reply src/prefect/flows.py:164]: Sure thing, the names are now confusing me too.

**Comment by rkscodes** [comment]
Updated with suggested changes, have a look.

File: src/prefect/flows.py [hunk 5]
```diff
@@ -158,6 +159,26 @@ def __init__(
             List[Callable[[schemas.core.Flow, FlowRun, State], None]]
         ] = None,
     ):
+        # Validate if hook passed is list and contains callables
+        hook_types = [on_completion, on_failure, on_cancellation, on_crashed]
```
Comment (serinamarie) [inline src/prefect/flows.py:163]: I hate to be a stickler *but* since we're doing duck typing below it seems a tad confusing to have `_types` in this variable (to me). I might suggest `hook_categories` or even `hook_kinds`🤷‍♀️

File: src/prefect/flows.py [hunk 6]
```diff
@@ -158,6 +159,26 @@ def __init__(
...
+                except TypeError:
+                    raise TypeError(
+                        f"Expected iterable for '{hook_name}'; got"
+                        f" {type(hook).__name__} instead."
+                    )
+
+                for hook_item in hook:
+                    if not callable(hook_item):
+                        raise TypeError(
+                            f"Expected callables in {hook}; got"
```
Comment (serinamarie) [inline src/prefect/flows.py:178]: I believe it would be more informative to include `hook_name` in this error message, to identify which hook category contains the non-callable item.

File: src/prefect/flows.py [hunk 7]
```diff
@@ -158,6 +159,26 @@ def __init__(
...
+            if hook is not None:
+                try:
+                    hook = list(hook)
+                except TypeError:
+                    raise TypeError(
+                        f"Expected iterable for '{hook_name}'; got"
+                        f" {type(hook).__name__} instead."
+                    )
+
+                for hook_item in hook:
```
Comment (serinamarie) [inline src/prefect/flows.py:175]: `for hook in hooks`: 
might be more intuitive (since right now `hook` is a list)

File: src/prefect/tasks.py [hunk 8]
```diff
@@ -51,7 +51,8 @@
     from prefect.context import TaskRunContext
 
 
-T = TypeVar("T")  # Generic type var for capturing the inner return type of async funcs
+# Generic type var for capturing the inner return type of async funcs
+T = TypeVar("T")
```
Comment (serinamarie) [inline src/prefect/tasks.py:55]: [As Zanie mentioned](https://github.com/PrefectHQ/prefect/pull/9692#discussion_r1213192612), this might have been accidentally modified?

**Review by serinamarie — CHANGES_REQUESTED** [review]
Hi @rkscodes, this is looking good! [As Zanie mentioned](https://github.com/PrefectHQ/prefect/pull/9692#pullrequestreview-1455539125), we'll want tests for this.

File: src/prefect/flows.py (hunk 5 above)
Comment (rkscodes) [reply src/prefect/flows.py:163]: Yeah sure, changed it to `hook_categories`

File: src/prefect/flows.py (hunk 7 above)
Comment (rkscodes) [reply src/prefect/flows.py:175]: Sure, `hook` is list, so I think I would have to rename it to `hooks` and then I can make it `for hook in hooks`

File: src/prefect/tasks.py (hunk 8 above)
Comment (rkscodes) [reply src/prefect/tasks.py:55]: Yup, reverting this too

File: src/prefect/flows.py (hunk 6 above)
Comment (rkscodes) [reply src/prefect/flows.py:178]: Noted

**Comment by rkscodes** [comment]
Updated with all suggested changes. I am really new to tests never written one should I attempt it ?

**Comment by serinamarie** [comment]
> Updated with all suggested changes. I am really new to tests never written one should I attempt it ?

Yeah, go ahead with what Zanie suggested regarding tests!

**Comment by rkscodes** [comment]
Added the test-cases, have a look!

**Review by serinamarie — COMMENTED** [review]
These first few tests are looking good! A few things:

1. Renaming these `test()` functions to `<hook_category>_hook` would be more comprehendible, e.g. `def completion_hook():`.

2. Adding tests for an empty iterable would make sense, e.g. `@flow(on_completion=[])`.

3. Adding tests for mixed iterables would also be ideal, e.g. `@flow(on_completion=[failure_hook, "non-callable"])`.

Let me know if you've got any questions! 🚀

**Comment by rkscodes** [comment]
I am bit unsure about point 1. So this fun  `def test_on_completion_hooks_is_iterable()` should be renamed to  
`def completion_hook_is_iterable()` ?

**Comment by rkscodes** [comment]
Yeah I am not able to get point 1. Could you give me an example let's say for `def test_on_completion_hooks_is_iterable()` what should I rename it to 

My guess : `def completion_hook_is_iterable()` 

Or are you suggesting to merge `... is_iterable` and `... is_callable` functions into single fun named `def completion_hook()`

**Comment by rkscodes** [comment]
Weird my last commit is not showing in this PR

**Comment by madkinsz** [comment]
@rkscodes I think GitHub is having a service disruption

**Comment by rkscodes** [comment]
Last commit is now visible, please have a look and let me know changes required if any.

File: tests/test_flows.py [hunk 9]
```diff
@@ -2364,6 +2403,44 @@ def my_flow():
...
+
+            @flow(on_failure=failure_hook)
+            def flow1():
+                pass
+
+        @flow(on_failure=[])
+        def flow2():
+            pass
+
+        assert flow2.on_failure == []
```
Comment (serinamarie) [inline tests/test_flows.py:2422]: I would expect this to raise. Also, it's probably best to move this to a separate test.

File: tests/test_flows.py [hunk 10]
```diff
@@ -2364,6 +2403,44 @@ def my_flow():
 
 
 class TestFlowHooksOnFailure:
+    def test_on_failure_hooks_is_iterable(self):
```
Comment (serinamarie) [inline tests/test_flows.py:2406]: I think it might make sense to name the tests like so:
`test_noniterable_hook_raises()`
`test_empty_hook_list_raises()`
`test_noncallable_hook_raises()`

File: tests/test_flows.py [hunk 11]
```diff
@@ -2364,6 +2403,44 @@ def my_flow():
...
+            def flow1():
+                pass
+
+        with pytest.raises(
+            TypeError, match="Expected callables in 'on_failure'; got str instead."
+        ):
+
+            @flow(on_failure=[failure_hook, "test"])
+            def flow2():
+                pass
```
Comment (serinamarie) [inline tests/test_flows.py:2442]: I would move this to a separate test

File: tests/test_flows.py (hunk 9 above)
Comment (rkscodes) [reply tests/test_flows.py:2422]: I think default value for `on_failure` is `[ ]`, so isn't is allowed to have empty list, please help me understand.

Comment (rkscodes) [reply tests/test_flows.py:2422]: Got it, I assumed default value to be empty list which is not true it is `None`

Comment (rkscodes) [reply tests/test_flows.py:2422]: In `flow.py` and `tasks.py` we are not concerned about empty list as values, I think I need to update code there first.

Comment (serinamarie) [reply tests/test_flows.py]: Let's check for the empty list first:

```python
if hooks is not None:
  if not hooks:  # This is equivalent to checking if len(hooks) == 0 for lists
      raise ValueError(f"Empty list passed for '{hook_name}'")
  try:
      hooks = list(hooks)
```

Comment (rkscodes) [reply tests/test_flows.py]: Sure, it's done.

File: src/prefect/flows.py [hunk 12]
```diff
@@ -158,6 +159,28 @@ def __init__(
...
+        # Validate if hook passed is list and contains callables
+        hook_categories = [on_completion, on_failure, on_cancellation, on_crashed]
+        hook_names = ["on_completion", "on_failure", "on_cancellation", "on_crashed"]
+        for hooks, hook_name in zip(hook_categories, hook_names):
+            if hooks is not None:
+                if not hooks:
+                    raise ValueError(f"Empty list passed for '{hook_name}'")
+                try:
+                    hooks = list(hooks)
+                except TypeError:
```
Comment (serinamarie) [inline src/prefect/flows.py:171]: @rkscodes, this is looking great! The only thing I can think of is to try to avoid a dead end when raising this `TypeError.` It would be more helpful to show the user what is expected:

```python
            raise TypeError(
                f"Expected iterable for '{hook_name}'; got"
                f" {type(hooks).__name__} instead. Please provide a list of hooks to 
                f"{hook_name}. For example:"
                f"\n@flow(on_completion=[hook1, hook2])\ndef flow1():\npass"
            )
```

Comment (rkscodes) [reply src/prefect/flows.py:171]: Yeah that would be more helpful for the user. I will adding the same msg looks good to me.

File: src/prefect/flows.py [hunk 13]
```diff
@@ -158,6 +159,34 @@ def __init__(
...
+                    raise TypeError(
+                        f"Expected iterable for '{hook_name}'; got"
+                        f" {type(hooks).__name__} instead. Please provide a list of"
+                        f" hooks to '{hook_name}'. For"
+                        f" example:\n@flow({hook_name}=[hook1, hook2])\ndef"
+                        " flow1():\npass"
+                    )
```
Comment (serinamarie) [inline src/prefect/flows.py:178]:
```suggestion
                    raise TypeError(
                        f"Expected iterable for '{hook_name}'; got"
                        f" {type(hooks).__name__} instead. Please provide a list of"
                        f" hooks to '{hook_name}':\n\n"
                        f"@flow({hook_name}=[hook1, hook2])\ndef"
                        " my_flow():\n\tpass"
                    )
```
Let's add indentation here, it will look nicer when it prints out:

```bash
  File "/Users/bean/code/prefect/src/prefect/flows.py", line 172, in __init__
    raise TypeError(
TypeError: Expected iterable for 'on_completion'; got function instead. Please provide a list of hooks to 'on_completion':

@flow(on_completion=[hook1, hook2])
def my_flow():
        pass
```

Comment (rkscodes) [reply src/prefect/flows.py:178]: Yeah it looks great, on it.

**Comment by serinamarie** [comment]
@rito-sixt This looks good, but there are changes to 2 additional files from [4893a52](https://github.com/PrefectHQ/prefect/pull/9692/commits/4893a524f52ecd8b433982d393a748c9c056cfdb)

**Comment by rito-sixt** [comment]
> @rito-sixt This looks good, but there are changes to 2 additional files from [4893a52](https://github.com/PrefectHQ/prefect/pull/9692/commits/4893a524f52ecd8b433982d393a748c9c056cfdb)

Me? 🤔

**Comment by rkscodes** [comment]
Ooops no idea those got added, reverting it right now.

**Comment by rkscodes** [comment]
Alright, I am looking into it serinamarie.

**Comment by rkscodes** [comment]
@serinamarie  Earlier `[]` was allowed but now we have to change it, I think `None` makes sense now over there.

**Review by serinamarie — APPROVED** [review]
Thanks a lot for your contribution and for going through several rounds of review, @rkscodes!

**Comment by rkscodes** [comment]
It was really amazing, I learned a lot thanks for guiding 🎉
